File: core-rs/core/src/parsing/media_tokens.rs
//! Media token extraction from LLM output text.
//!
//! Ports `src/media/parse.ts:splitMediaFromOutput`.
//! Extracts `MEDIA: <url/path>` tokens from text while respecting
//! fenced code blocks, and detects `[[audio_as_voice]]` tags.

use serde::Serialize;

/// Result of media token parsing.
#[derive(Debug, Serialize)]
pub struct MediaParseResult {
    pub text: String,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub media_urls: Vec<String>,
    /// First media URL for backward compatibility.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub media_url: Option<String>,
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    pub audio_as_voice: bool,
}

/// Span of a fenced code block.
struct FenceSpan {
    start: usize,
    end: usize,
}

/// Parse fenced code block spans (`` ``` `` or `~~~`).
fn parse_fence_spans(text: &str) -> Vec<FenceSpan> {
    let mut spans = Vec::new();
    let lines: Vec<&str> = text.split('\n').collect();
    let mut in_fence = false;
    let mut fence_char = b'`';
    let mut fence_len = 0;
    let mut fence_start = 0;
    let mut offset = 0;

    for line in &lines {
        let trimmed = line.trim_start();
        if !in_fence {
            if let Some((ch, len)) = detect_fence_open(trimmed) {
                in_fence = true;
                fence_char = ch;
                fence_len = len;
                fence_start = offset;
            }
        } else if is_fence_close(trimmed, fence_char, fence_len) {
            spans.push(FenceSpan {
                start: fence_start,
                end: offset + line.len(),
            });
            in_fence = false;
        }
        offset += line.len() + 1; // +1 for newline
    }

    // If fence was never closed, extend to end.
    if in_fence {
        spans.push(FenceSpan {
            start: fence_start,
            end: text.len(),
        });
    }

    spans
}

fn detect_fence_open(trimmed: &str) -> Option<(u8, usize)> {
    let bytes = trimmed.as_bytes();
    if bytes.len() < 3 {
        return None;
    }
    let ch = bytes[0];
    if ch != b'`' && ch != b'~' {
        return None;
    }
    let mut count = 0;
    for &b in bytes {
        if b == ch {
            count += 1;
        } else {
            break;
        }
    }
    if count >= 3 {
        Some((ch, count))
    } else {
        None
    }
}

fn is_fence_close(trimmed: &str, fence_char: u8, fence_len: usize) -> bool {
    let bytes = trimmed.as_bytes();
    if bytes.len() < fence_len {
        return false;
    }
    let mut count = 0;
    for &b in bytes {
        if b == fence_char {
            count += 1;
        } else if b == b' ' || b == b'\t' {
            // trailing whitespace is OK
            break;
        } else {
            return false;
        }
    }
    count >= fence_len
}

fn is_inside_fence(spans: &[FenceSpan], offset: usize) -> bool {
    spans.iter().any(|s| offset >= s.start && offset < s.end)
}

/// Check if candidate looks like a valid media source (URL or local path).
fn is_valid_media(candidate: &str) -> bool {
    if candidate.is_empty() || candidate.len() > 4096 {
        return false;
    }
    // No whitespace allowed in simple mode.
    if candidate.chars().any(char::is_whitespace) {
        return false;
    }
    is_valid_media_core(candidate) || is_bare_filename(candidate)
}

/// Core path/URL validation without whitespace restriction.
fn is_valid_media_core(candidate: &str) -> bool {
    // HTTP(S) URL.
    if candidate.starts_with("http://") || candidate.starts_with("https://") {
        return true;
    }
    // Local path patterns.
    if candidate.starts_with('/')
        || candidate.starts_with("./")
        || candidate.starts_with("../")
        || candidate.starts_with('~')
    {
        return true;
    }
    // Windows drive letter.
    let bytes = candidate.as_bytes();
    if bytes.len() >= 3
        && bytes[0].is_ascii_alphabetic()
        && bytes[1] == b':'
        && (bytes[2] == b'\\' || bytes[2] == b'/')
    {
        return true;
    }
    // UNC path.
    if candidate.starts_with("\\\\") {
        return true;
    }
    false
}

/// Check if a candidate is a bare filename with a file extension (1-10 chars).
/// E.g., "image.png", "recording.m4a"
fn is_bare_filename(candidate: &str) -> bool {
    if candidate.is_empty() || candidate.len() > 260 {
        return false;
    }
    if let Some(dot_pos) = candidate.rfind('.') {
        let ext_len = candidate.len() - dot_pos - 1;
        if (1..=10).contains(&ext_len) {
            let name = &candidate[..dot_pos];
            // Name must be non-empty and contain no path separators
            return !name.is_empty() && !name.contains('/') && !name.contains('\\');
        }
    }
    false
}

/// Try to extract a quoted string payload from the text.
/// Supports double quotes, single quotes.
/// Returns the unquoted content if found.
fn try_unwrap_quoted(payload: &str) -> Option<&str> {
    let trimmed = payload.trim();
    if trimmed.len() < 2 {
        return None;
    }
    let first = trimmed.as_bytes()[0];
    let last = trimmed.as_bytes()[trimmed.len() - 1];
    if (first == b'"' && last == b'"') || (first == b'\'' && last == b'\'') {
        Some(&trimmed[1..trimmed.len() - 1])
    } else {
        None
    }
}

/// Clean wrapping quotes/brackets from a candidate.
fn clean_candidate(raw: &str) -> &str {
    let mut s = raw;
    // Strip leading punctuation.
    while !s.is_empty() {
        let first = s.as_bytes()[0];
        if matches!(first, b'`' | b'"' | b'\'' | b'[' | b'{' | b'(') {
            s = &s[1..];
        } else {
            break;
        }
    }
    // Strip trailing punctuation.
    while !s.is_empty() {
        let last = s.as_bytes()[s.len() - 1];
        if matches!(
            last,
            b'`' | b'"' | b'\'' | b'\\' | b'}' | b')' | b']' | b','
        ) {
            s = &s[..s.len() - 1];
        } else {
            break;
        }
    }
    s
}

/// Normalize `file://` prefix.
fn normalize_media_source(src: &str) -> String {
    if let Some(rest) = src.strip_prefix("file://") {
        rest.to_string()
    } else {
        src.to_string()
    }
}

/// A parsed inline directive like `[[key]]` or `[[key=value]]`.
#[derive(Debug)]
struct InlineDirective {
    key: String,
}

/// Parse and strip all `[[...]]` inline directives from text.
/// Returns the cleaned text and a list of parsed directives.
fn strip_inline_directives(text: &str) -> (String, Vec<InlineDirective>) {
    let mut directives = Vec::new();
    let mut result = String::with_capacity(text.len());
    let mut rest = text;

    while let Some(start) = rest.find("[[") {
        if let Some(end) = rest[start + 2..].find("]]") {
            let inner = &rest[start + 2..start + 2 + end];
            // Parse key=value or just key
            let directive = if let Some(eq_pos) = inner.find('=') {
                InlineDirective {
                    key: inner[..eq_pos].trim().to_string(),
                }
            } else {
                InlineDirective {
                    key: inner.trim().to_string(),
                }
            };
            result.push_str(&rest[..start]);
            directives.push(directive);
            rest = &rest[start + 2 + end + 2..];
        } else {
            // No closing ]] found — keep the rest as-is
            result.push_str(rest);
            rest = "";
            break;
        }
    }
    result.push_str(rest);

    (result, directives)
}

/// Strip directives and detect `audio_as_voice`.
fn strip_audio_tag(text: &str) -> (String, bool) {
    let (cleaned, directives) = strip_inline_directives(text);
    let audio_as_voice = directives
        .iter()
        .any(|d| d.key == "audio_as_voice" || d.key == "voice");
    (cleaned, audio_as_voice)
}

/// Extract `MEDIA:` tokens from text output.
///
/// Returns cleaned text with MEDIA lines removed, extracted media URLs,
/// and whether `[[audio_as_voice]]` was present.
pub fn split_media_from_output(raw: &str) -> MediaParseResult {
    let trimmed_raw = raw.trim_end();
    if trimmed_raw.trim().is_empty() {
        return MediaParseResult {
            text: String::new(),
            media_urls: Vec::new(),
            media_url: None,
            audio_as_voice: false,
        };
    }

    let has_media_token = trimmed_raw
        .as_bytes()
        .windows(6)
        .any(|w| w.eq_ignore_ascii_case(b"media:"));
    let has_audio_tag = trimmed_raw.contains("[[");

    if !has_media_token && !has_audio_tag {
        return MediaParseResult {
            text: trimmed_raw.to_string(),
            media_urls: Vec::new(),
            media_url: None,
            audio_as_voice: false,
        };
    }

    let has_fence_markers = trimmed_raw.contains("```") || trimmed_raw.contains("~~~");
    let fence_spans = if has_fence_markers {
        parse_fence_spans(trimmed_raw)
    } else {
        Vec::new()
    };

    let mut media: Vec<String> = Vec::new();
    let mut found_media_token = false;
    let lines: Vec<&str> = trimmed_raw.split('\n').collect();
    let mut kept_lines: Vec<String> = Vec::new();
    let mut line_offset = 0;

    for line in &lines {
        // Skip MEDIA extraction inside fenced code blocks.
        if has_fence_markers && is_inside_fence(&fence_spans, line_offset) {
            kept_lines.push(line.to_string());
            line_offset += line.len() + 1;
            continue;
        }

        let trimmed_start = line.trim_start();
        if !trimmed_start.starts_with("MEDIA:") {
            kept_lines.push(line.to_string());
            line_offset += line.len() + 1;
            continue;
        }

        // Extract the payload after "MEDIA:".
        let Some(media_idx) = trimmed_start.find("MEDIA:") else {
            // Verified by starts_with above; unreachable in practice.
            kept_lines.push(line.to_string());
            line_offset += line.len() + 1;
            continue;
        };
        let payload = &trimmed_start[media_idx + 6..];
        let payload = payload.trim();
        // Strip optional wrapping backtick.
        let payload = payload.strip_prefix('`').unwrap_or(payload);
        let payload = payload.strip_suffix('`').unwrap_or(payload);
        let payload = payload.trim();

        if payload.is_empty() {
            kept_lines.push(line.to_string());
            line_offset += line.len() + 1;
            continue;
        }

        // Stage 1: Try unwrapping quoted payload (e.g., MEDIA: "path with spaces.mp3").
        let mut any_valid = false;
        if let Some(unquoted) = try_unwrap_quoted(payload) {
            let candidate = clean_candidate(unquoted);
            let normalized = normalize_media_source(candidate);
            if is_valid_media_allow_spaces(&normalized) || is_bare_filename(&normalized) {
                media.push(normalized);
                any_valid = true;
                found_media_token = true;
            }
        }

        // Stage 2: Try each space-separated part.
        if !any_valid {
            for part in payload.split_whitespace() {
                let candidate = clean_candidate(part);
                let normalized = normalize_media_source(candidate);
                if is_valid_media(&normalized) {
                    media.push(normalized);
                    any_valid = true;
                    found_media_token = true;
                }
            }
        }

        // Stage 3: Fallback — try entire payload as a single path (may contain spaces).
        if !any_valid {
            let candidate = clean_candidate(payload);
            let normalized = normalize_media_source(candidate);
            if is_valid_media_allow_spaces(&normalized) || is_bare_filename(&normalized) {
                media.push(normalized);
                found_media_token = true;
                any_valid = true;
            }
        }

        if !any_valid {
            // Check if it looks like a local path — strip even if invalid.
            let candidate = clean_candidate(payload);
            if is_likely_local_path(candidate) {
                found_media_token = true;
                // Drop the line (don't keep it).
            } else {
                kept_lines.push(line.to_string());
            }
        }
        // If valid media found, the MEDIA line is stripped (not added to kept_lines).

        line_offset += line.len() + 1;
    }

    let mut cleaned_text = kept_lines.join("\n");
    // Collapse multiple spaces and blank lines.
    cleaned_text = collapse_whitespace(&cleaned_text);

    // Strip [[audio_as_voice]] tag.
    let (cleaned_text, audio_as_voice) = strip_audio_tag(&cleaned_text);
    let cleaned_text = if audio_as_voice {
        collapse_whitespace(&cleaned_text)
    } else {
        cleaned_text
    };
    let cleaned_text = cleaned_text.trim().to_string();

    if media.is_empty() {
        return MediaParseResult {
            text: if found_media_token || audio_as_voice {
                cleaned_text
            } else {
                trimmed_raw.to_string()
            },
            media_urls: Vec::new(),
            media_url: None,
            audio_as_voice,
        };
    }

    let media_url = media.first().cloned();
    MediaParseResult {
        text: cleaned_text,
        media_urls: media,
        media_url,
        audio_as_voice,
    }
}

fn is_valid_media_allow_spaces(candidate: &str) -> bool {
    if candidate.is_empty() || candidate.len() > 4096 {
        return false;
    }
    is_valid_media_core(candidate)
}

fn is_likely_local_path(candidate: &str) -> bool {
    candidate.starts_with('/')
        || candidate.starts_with("./")
        || candidate.starts_with("../")
        || candidate.starts_with('~')
        || candidate.starts_with("file://")
        || candidate.starts_with("\\\\")
}

/// Single-pass whitespace normalization:
/// - Trim trailing whitespace on each line
/// - Collapse consecutive blank lines into one newline
/// - Collapse consecutive spaces/tabs into a single space
fn collapse_whitespace(input: &str) -> String {
    let mut out = String::with_capacity(input.len());
    // trailing_ws tracks spaces/tabs that may be trailing (not yet flushed).
    let mut trailing_ws: usize = 0;
    let mut newline_count: u32 = 0;
    let mut prev_space = false;

    for ch in input.chars() {
        match ch {
            '\n' => {
                // Drop any trailing whitespace before the newline.
                trailing_ws = 0;
                prev_space = false;
                newline_count += 1;
                if newline_count <= 1 {
                    out.push('\n');
                }
            }
            ' ' | '\t' => {
                newline_count = 0;
                // Buffer this as potential trailing whitespace.
                if !prev_space {
                    trailing_ws += 1;
                }
                prev_space = true;
            }
            _ => {
                // Flush buffered whitespace as a single space.
                if trailing_ws > 0 || prev_space {
                    out.push(' ');
                    trailing_ws = 0;
                }
                prev_space = false;
                newline_count = 0;
                out.push(ch);
            }
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_input() {
        let result = split_media_from_output("");
        assert_eq!(result.text, "");
        assert!(result.media_urls.is_empty());
    }

    #[test]
    fn no_media_tokens() {
        let result = split_media_from_output("Hello world, no media here.");
        assert_eq!(result.text, "Hello world, no media here.");
        assert!(result.media_urls.is_empty());
    }

    #[test]
    fn single_url_token() {
        let result =
            split_media_from_output("Here is an image\nMEDIA: https://example.com/image.png");
        assert_eq!(result.media_urls, vec!["https://example.com/image.png"]);
        assert!(result.media_url.is_some());
        assert!(result.text.contains("Here is an image"));
        assert!(!result.text.contains("MEDIA:"));
    }

    #[test]
    fn local_path_token() {
        let result = split_media_from_output("MEDIA: /tmp/output.wav\nDone.");
        assert_eq!(result.media_urls, vec!["/tmp/output.wav"]);
        assert!(result.text.contains("Done."));
    }

    #[test]
    fn file_protocol_normalized() {
        let result = split_media_from_output("MEDIA: file:///tmp/audio.mp3");
        assert_eq!(result.media_urls, vec!["/tmp/audio.mp3"]);
    }

    #[test]
    fn inside_fence_ignored() {
        let input = "text\n```\nMEDIA: https://example.com/skip.png\n```\nMEDIA: https://example.com/keep.png";
        let result = split_media_from_output(input);
        assert_eq!(result.media_urls.len(), 1);
        assert_eq!(result.media_urls[0], "https://example.com/keep.png");
        // The fenced MEDIA line should be kept as text.
        assert!(result.text.contains("MEDIA: https://example.com/skip.png"));
    }

    #[test]
    fn audio_as_voice_tag() {
        let result = split_media_from_output("Hello [[audio_as_voice]]\nMEDIA: /tmp/voice.wav");
        assert!(result.audio_as_voice);
        assert!(!result.text.contains("[[audio_as_voice]]"));
    }

    #[test]
    fn multiple_media_urls() {
        let input = "MEDIA: https://a.com/1.png\ntext\nMEDIA: https://b.com/2.png";
        let result = split_media_from_output(input);
        assert_eq!(result.media_urls.len(), 2);
    }

    #[test]
    fn backtick_wrapped() {
        let result = split_media_from_output("MEDIA: `https://example.com/img.png`");
        assert_eq!(result.media_urls, vec!["https://example.com/img.png"]);
    }

    #[test]
    fn quoted_path_with_spaces() {
        let result = split_media_from_output(r#"MEDIA: "/tmp/my file with spaces.mp3""#);
        assert_eq!(result.media_urls, vec!["/tmp/my file with spaces.mp3"]);
    }

    #[test]
    fn bare_filename() {
        let result = split_media_from_output("MEDIA: image.png");
        assert_eq!(result.media_urls, vec!["image.png"]);
    }

    #[test]
    fn bare_filename_with_extension() {
        let result = split_media_from_output("MEDIA: recording.m4a");
        assert_eq!(result.media_urls, vec!["recording.m4a"]);
    }

    #[test]
    fn directive_key_value() {
        let result = split_media_from_output(
            "Hello [[audio_as_voice]] [[format=wav]]\nMEDIA: /tmp/voice.wav",
        );
        assert!(result.audio_as_voice);
        assert!(!result.text.contains("[["));
    }

    #[test]
    fn directive_unclosed_bracket() {
        let result = split_media_from_output("Hello [[ not closed");
        assert_eq!(result.text, "Hello [[ not closed");
        assert!(!result.audio_as_voice);
    }

    #[test]
    fn voice_tag_alias() {
        let result = split_media_from_output("Hello [[voice]]\nMEDIA: /tmp/voice.wav");
        assert!(result.audio_as_voice);
        assert!(!result.text.contains("[[voice]]"));
    }
}

File: gateway-go/internal/core/coreparsing/mediatokens/mediatokens.go
// Package mediatokens extracts MEDIA: tokens from LLM output text.
//
// Ported from core-rs/core/src/parsing/media_tokens.rs.
// Extracts MEDIA: <url/path> tokens from text while respecting fenced code
// blocks, and detects [[audio_as_voice]] tags.
package mediatokens

import (
	"strings"
)

// Result of media token parsing.
type Result struct {
	Text         string   // Cleaned text (MEDIA lines removed).
	MediaURLs    []string // Extracted URLs/paths.
	AudioAsVoice bool     // [[audio_as_voice]] detected.
}

// Parse extracts MEDIA: tokens from text output.
// Returns cleaned text with MEDIA lines removed, extracted media URLs,
// and whether [[audio_as_voice]] was present.
func Parse(raw string) Result {
	trimmedRaw := strings.TrimRight(raw, " \t\n\r")
	if strings.TrimSpace(trimmedRaw) == "" {
		return Result{}
	}

	hasMediaToken := containsIgnoreCase(trimmedRaw, "media:")
	hasAudioTag := strings.Contains(trimmedRaw, "[[")

	if !hasMediaToken && !hasAudioTag {
		return Result{Text: trimmedRaw}
	}

	hasFenceMarkers := strings.Contains(trimmedRaw, "```") || strings.Contains(trimmedRaw, "~~~")
	var fenceSpans []fenceSpan
	if hasFenceMarkers {
		fenceSpans = parseFenceSpans(trimmedRaw)
	}

	lines := strings.Split(trimmedRaw, "\n")
	var keptLines []string
	var media []string
	foundMediaToken := false
	lineOffset := 0

	for _, line := range lines {
		// Skip MEDIA extraction inside fenced code blocks.
		if hasFenceMarkers && isInsideFence(fenceSpans, lineOffset) {
			keptLines = append(keptLines, line)
			lineOffset += len(line) + 1
			continue
		}

		trimmedStart := strings.TrimLeft(line, " \t")
		if !strings.HasPrefix(trimmedStart, "MEDIA:") {
			keptLines = append(keptLines, line)
			lineOffset += len(line) + 1
			continue
		}

		// Extract payload after "MEDIA:".
		payload := strings.TrimSpace(trimmedStart[6:])
		// Strip optional wrapping backtick.
		payload = strings.TrimPrefix(payload, "`")
		payload = strings.TrimSuffix(payload, "`")
		payload = strings.TrimSpace(payload)

		if payload == "" {
			keptLines = append(keptLines, line)
			lineOffset += len(line) + 1
			continue
		}

		anyValid := false

		// Stage 1: Try unwrapping quoted payload.
		if unquoted, ok := tryUnwrapQuoted(payload); ok {
			candidate := cleanCandidate(unquoted)
			normalized := normalizeMediaSource(candidate)
			if isValidMediaAllowSpaces(normalized) || isBareFilename(normalized) {
				media = append(media, normalized)
				anyValid = true
				foundMediaToken = true
			}
		}

		// Stage 2: Try each space-separated part.
		if !anyValid {
			for _, part := range strings.Fields(payload) {
				candidate := cleanCandidate(part)
				normalized := normalizeMediaSource(candidate)
				if isValidMedia(normalized) {
					media = append(media, normalized)
					anyValid = true
					foundMediaToken = true
				}
			}
		}

		// Stage 3: Fallback — try entire payload as a single path.
		if !anyValid {
			candidate := cleanCandidate(payload)
			normalized := normalizeMediaSource(candidate)
			if isValidMediaAllowSpaces(normalized) || isBareFilename(normalized) {
				media = append(media, normalized)
				foundMediaToken = true
				anyValid = true
			}
		}

		if !anyValid {
			candidate := cleanCandidate(payload)
			if isLikelyLocalPath(candidate) {
				foundMediaToken = true
				// Drop the line.
			} else {
				keptLines = append(keptLines, line)
			}
		}

		lineOffset += len(line) + 1
	}

	cleanedText := strings.Join(keptLines, "\n")
	cleanedText = collapseWhitespace(cleanedText)

	// Strip [[audio_as_voice]] tag.
	cleanedText, audioAsVoice := stripAudioTag(cleanedText)
	if audioAsVoice {
		cleanedText = collapseWhitespace(cleanedText)
	}
	cleanedText = strings.TrimSpace(cleanedText)

	if len(media) == 0 {
		text := trimmedRaw
		if foundMediaToken || audioAsVoice {
			text = cleanedText
		}
		return Result{Text: text, AudioAsVoice: audioAsVoice}
	}

	return Result{
		Text:         cleanedText,
		MediaURLs:    media,
		AudioAsVoice: audioAsVoice,
	}
}

// containsIgnoreCase checks if s contains substr (case-insensitive).
func containsIgnoreCase(s, substr string) bool {
	sLen := len(s)
	subLen := len(substr)
	if subLen > sLen {
		return false
	}
	for i := 0; i <= sLen-subLen; i++ {
		match := true
		for j := 0; j < subLen; j++ {
			sc := s[i+j]
			tc := substr[j]
			if sc >= 'A' && sc <= 'Z' {
				sc += 'a' - 'A'
			}
			if tc >= 'A' && tc <= 'Z' {
				tc += 'a' - 'A'
			}
			if sc != tc {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}

// fenceSpan marks a fenced code block region.
type fenceSpan struct {
	start, end int
}

func parseFenceSpans(text string) []fenceSpan {
	var spans []fenceSpan
	lines := strings.Split(text, "\n")
	inFence := false
	var fenceChar byte
	fenceLen := 0
	fenceStart := 0
	offset := 0

	for _, line := range lines {
		trimmed := strings.TrimLeft(line, " \t")
		if !inFence {
			if ch, n := detectFenceOpen(trimmed); n >= 3 {
				inFence = true
				fenceChar = ch
				fenceLen = n
				fenceStart = offset
			}
		} else if isFenceClose(trimmed, fenceChar, fenceLen) {
			spans = append(spans, fenceSpan{start: fenceStart, end: offset + len(line)})
			inFence = false
		}
		offset += len(line) + 1
	}

	// If fence was never closed, extend to end.
	if inFence {
		spans = append(spans, fenceSpan{start: fenceStart, end: len(text)})
	}
	return spans
}

func detectFenceOpen(trimmed string) (byte, int) {
	if len(trimmed) < 3 {
		return 0, 0
	}
	ch := trimmed[0]
	if ch != '`' && ch != '~' {
		return 0, 0
	}
	count := 0
	for i := 0; i < len(trimmed); i++ {
		if trimmed[i] == ch {
			count++
		} else {
			break
		}
	}
	if count >= 3 {
		return ch, count
	}
	return 0, 0
}

func isFenceClose(trimmed string, fenceChar byte, fenceLen int) bool {
	if len(trimmed) < fenceLen {
		return false
	}
	count := 0
	for i := 0; i < len(trimmed); i++ {
		if trimmed[i] == fenceChar {
			count++
		} else if trimmed[i] == ' ' || trimmed[i] == '\t' {
			break
		} else {
			return false
		}
	}
	return count >= fenceLen
}

func isInsideFence(spans []fenceSpan, offset int) bool {
	for _, s := range spans {
		if offset >= s.start && offset < s.end {
			return true
		}
	}
	return false
}

// isValidMedia checks if candidate looks like a valid media source (no whitespace).
func isValidMedia(candidate string) bool {
	if candidate == "" || len(candidate) > 4096 {
		return false
	}
	if strings.ContainsAny(candidate, " \t\n\r") {
		return false
	}
	return isValidMediaCore(candidate) || isBareFilename(candidate)
}

// isValidMediaAllowSpaces is like isValidMedia but allows whitespace.
func isValidMediaAllowSpaces(candidate string) bool {
	if candidate == "" || len(candidate) > 4096 {
		return false
	}
	return isValidMediaCore(candidate)
}

// isValidMediaCore checks URL or local path patterns.
func isValidMediaCore(candidate string) bool {
	if strings.HasPrefix(candidate, "http://") || strings.HasPrefix(candidate, "https://") {
		return true
	}
	if strings.HasPrefix(candidate, "/") ||
		strings.HasPrefix(candidate, "./") ||
		strings.HasPrefix(candidate, "../") ||
		strings.HasPrefix(candidate, "~") {
		return true
	}
	// Windows drive letter.
	if len(candidate) >= 3 &&
		isASCIIAlpha(candidate[0]) &&
		candidate[1] == ':' &&
		(candidate[2] == '\\' || candidate[2] == '/') {
		return true
	}
	// UNC path.
	if strings.HasPrefix(candidate, "\\\\") {
		return true
	}
	return false
}

// isBareFilename checks if candidate is a bare filename with extension (1-10 chars).
func isBareFilename(candidate string) bool {
	if candidate == "" || len(candidate) > 260 {
		return false
	}
	dotPos := strings.LastIndexByte(candidate, '.')
	if dotPos < 0 {
		return false
	}
	extLen := len(candidate) - dotPos - 1
	if extLen < 1 || extLen > 10 {
		return false
	}
	name := candidate[:dotPos]
	return name != "" && !strings.ContainsAny(name, "/\\")
}

func isASCIIAlpha(b byte) bool {
	return (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z')
}

// tryUnwrapQuoted tries to extract content from a quoted string.
func tryUnwrapQuoted(payload string) (string, bool) {
	trimmed := strings.TrimSpace(payload)
	if len(trimmed) < 2 {
		return "", false
	}
	first := trimmed[0]
	last := trimmed[len(trimmed)-1]
	if (first == '"' && last == '"') || (first == '\'' && last == '\'') {
		return trimmed[1 : len(trimmed)-1], true
	}
	return "", false
}

// cleanCandidate strips leading/trailing wrapping chars.
func cleanCandidate(raw string) string {
	s := raw
	for len(s) > 0 {
		switch s[0] {
		case '`', '"', '\'', '[', '{', '(':
			s = s[1:]
		default:
			goto doneLeading
		}
	}
doneLeading:
	for len(s) > 0 {
		switch s[len(s)-1] {
		case '`', '"', '\'', '\\', '}', ')', ']', ',':
			s = s[:len(s)-1]
		default:
			return s
		}
	}
	return s
}

// normalizeMediaSource strips file:// prefix.
func normalizeMediaSource(src string) string {
	if rest, ok := strings.CutPrefix(src, "file://"); ok {
		return rest
	}
	return src
}

func isLikelyLocalPath(candidate string) bool {
	return strings.HasPrefix(candidate, "/") ||
		strings.HasPrefix(candidate, "./") ||
		strings.HasPrefix(candidate, "../") ||
		strings.HasPrefix(candidate, "~") ||
		strings.HasPrefix(candidate, "file://") ||
		strings.HasPrefix(candidate, "\\\\")
}

// stripAudioTag strips [[audio_as_voice]] and [[voice]] directives.
func stripAudioTag(text string) (string, bool) {
	cleaned, directives := stripInlineDirectives(text)
	audioAsVoice := false
	for _, d := range directives {
		if d == "audio_as_voice" || d == "voice" {
			audioAsVoice = true
		}
	}
	return cleaned, audioAsVoice
}

// stripInlineDirectives parses and strips all [[...]] inline directives.
// Returns cleaned text and a list of directive keys.
func stripInlineDirectives(text string) (string, []string) {
	var keys []string
	var result strings.Builder
	result.Grow(len(text))
	rest := text

	for {
		start := strings.Index(rest, "[[")
		if start < 0 {
			result.WriteString(rest)
			break
		}
		endIdx := strings.Index(rest[start+2:], "]]")
		if endIdx < 0 {
			// No closing ]] — keep rest as-is.
			result.WriteString(rest)
			break
		}
		inner := rest[start+2 : start+2+endIdx]
		// Parse key=value or just key.
		key := inner
		if eqPos := strings.IndexByte(inner, '='); eqPos >= 0 {
			key = inner[:eqPos]
		}
		key = strings.TrimSpace(key)
		result.WriteString(rest[:start])
		keys = append(keys, key)
		rest = rest[start+2+endIdx+2:]
	}

	return result.String(), keys
}

// collapseWhitespace normalizes whitespace:
// - Trims trailing whitespace on each line
// - Collapses consecutive blank lines into one newline
// - Collapses consecutive spaces/tabs into a single space
func collapseWhitespace(input string) string {
	var out strings.Builder
	out.Grow(len(input))
	trailingWS := 0
	newlineCount := 0
	prevSpace := false

	for _, ch := range input {
		switch ch {
		case '\n':
			trailingWS = 0
			prevSpace = false
			newlineCount++
			if newlineCount <= 1 {
				out.WriteByte('\n')
			}
		case ' ', '\t':
			newlineCount = 0
			if !prevSpace {
				trailingWS++
			}
			prevSpace = true
		default:
			if trailingWS > 0 || prevSpace {
				out.WriteByte(' ')
				trailingWS = 0
			}
			prevSpace = false
			newlineCount = 0
			out.WriteRune(ch)
		}
	}
	return out.String()
}
